### PR TITLE
Add PyPI Observation API integration

### PR DESCRIPTION
**Pull request by Robin5605**

### CODE REVIEW

File: src/dragonfly_reporter/models.py [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+
+class Observation(BaseModel):
+    kind: ObservationKind
+    summary: str
+    inspector_url: Optional[str]
+    extra: dict[str, Any] = {}
+
+    @model_validator(mode="after")
+    def model_validator(self: Observation) -> Observation:
+        if self.kind == ObservationKind.Malware:
```
Comment (jonathan-d-zhang) [inline src/dragonfly_reporter/models.py:24]: Maybe worth testing this condition

File: src/dragonfly_reporter/observations.py [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+from dragonfly_reporter.models import Observation
+
+
+async def send_observation(
+    project_name: str, observation: Observation, *, http_client: httpx.AsyncClient
+):
+    url = f"/danger-api/projects/{project_name}/observations"
+    json = jsonable_encoder(observation)
+
+    await http_client.post(url, json=json)
```
Comment (jonathan-d-zhang) [inline src/dragonfly_reporter/observations.py:15]: Technically not a url
```suggestion
    path = f"/danger-api/projects/{project_name}/observations"
    json = jsonable_encoder(observation)

    await http_client.post(path, json=json)
```

File: src/dragonfly_reporter/http_client.py [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+        self, request: httpx.Request
+    ) -> Generator[httpx.Request, httpx.Response, None]:
+        request.headers["Authorization"] = f"Bearer {self.token}"
+        yield request
+
+
+@cache
+def http_client() -> httpx.AsyncClient:
+    auth = BearerAuthentication(token=reporter_settings.observation_api_token)
+    http_client = httpx.AsyncClient(auth=auth, base_url=BASE_URL)
```
Comment (jonathan-d-zhang) [inline src/dragonfly_reporter/http_client.py:27]: This change means we aren't using context managers anymore, is that fine?

File: src/dragonfly_reporter/models.py [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+            assert (
```
Comment (AbooMinister25) [inline src/dragonfly_reporter/models.py:25]: This is totally a nit, but why the assert here?

**Review by AbooMinister25 — APPROVED** [review]
LGTM for the most part.

File: src/dragonfly_reporter/models.py (hunk 4 above)
Comment (Robin5605) [reply src/dragonfly_reporter/models.py:25]: It is required for Pydantic custom validators to work: https://docs.pydantic.dev/latest/concepts/validators/#field-validators

**Review by jonathan-d-zhang — CHANGES_REQUESTED** [review]

File: src/dragonfly_reporter/models.py (hunk 4 above)
Comment (AbooMinister25) [reply src/dragonfly_reporter/models.py:25]: gotcha

File: src/dragonfly_reporter/http_client.py (hunk 3 above)
Comment (Robin5605) [reply src/dragonfly_reporter/http_client.py:27]: Yes, though you do bring up a good point. Currently we create the HTTP client but don't close it anywhere. Perhaps this should be done in FastAPI's lifetime.

**Review by jonathan-d-zhang — APPROVED** [review]
:shipit:. Note that you'll have to modify the dockerfile if you've changed how to run the server.

**Review by jonathan-d-zhang — APPROVED** [review]
:shipit:

**Comment by Xithrius** [comment]
I forgor :skull:

**Pull request merged**